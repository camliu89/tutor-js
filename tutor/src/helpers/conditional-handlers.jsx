import React from 'react';
import invariant from 'invariant';
import { extend, last, first } from 'lodash';
import { Redirect } from 'react-router-dom';
import { asyncComponent } from './async-component';
import { CourseNotFoundWarning } from '../components/course-not-found-warning';
import Courses from '../models/courses-map';
import { OXMatchByRouter } from 'shared';

const StudentDashboard = asyncComponent(
  () => import('../screens/student-dashboard')
);

const getConditionalHandlers = (Router) => {
  const MatchForTutor = OXMatchByRouter(Router, null, 'TutorRouterMatch');

  const renderDashboard = (props) => {
    const { courseId } = props.params;

    extend(props, { courseId });
    const course = Courses.get(courseId);

    if (!course) {
      return <CourseNotFoundWarning />;
    }

    if (!props.match.isExact) {
      return (
        <MatchForTutor {...props} />
      );
    }

    if (course.isTeacher) {
      if (course.is_concept_coach) {
        return (
          <CCDashboard courseId={courseId} {...props} />
        );
      } else {
        return (
          <Redirect
            to={{
              pathname: Router.makePathname('viewTeacherDashboard', props.params),
              query: Router.currentQuery(),
            }} />
        );
      }
    } else {
      if (course.is_concept_coach) {
        return (
          <CCStudentRedirect courseId={courseId} />
        );
      } else {
        return <StudentDashboard {...props} />;
      }
    }
  };
  // eslint-disable-next-line react/prop-types
  const legacyReferenceBookRedirect = ({ params }) => {
    const parts = params.parts.split('/');
    const course = Courses.get(first(parts));
    invariant(course, `Did not find course for params '${params}'`);
    const chapterSection = parts.length > 1 ? last(parts) : 0;
    return (
      <Redirect
        to={{
          pathname: Router.makePathname(
            chapterSection ? 'viewReferenceBookSection' : 'viewReferenceBook',
            {
              ecosystemId: course.ecosystem_id,
              chapterSection,
            }),
          query: Router.currentQuery(),
        }}
      />
    );
  };

  return {
    dashboard() { return renderDashboard; },
    legacyReferenceBookRedirect() { return legacyReferenceBookRedirect; },
  };
};

export { getConditionalHandlers };
