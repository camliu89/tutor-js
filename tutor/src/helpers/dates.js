import moment from 'moment';
import 'twix';
import 'moment-timezone';
import { compact, reduce, min, max, map } from 'lodash';
import TimeHelper from './time';

export function getDay(oneMoment) {
  return moment(oneMoment)
    .startOf('day')
    .twix(moment(oneMoment).endOf('day'), { allDay: true });
}

export function getDurationFromMoments(listOfMoments) {
  return reduce(
    listOfMoments,
    (current, next) => current.union(getDay(next)),
    getDay(listOfMoments[0])
  );
}

export function findEarliest(dateThings) {
  return min(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}

export function findLatest(dateThings) {
  return max(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}


export function dateWithUnchangedTime(date, t) {
  if (moment.isMoment(date)) {
    date = date.format(TimeHelper.ISO_DATE_FORMAT);
  }
  return TimeHelper.getDateOnly(date);
}
