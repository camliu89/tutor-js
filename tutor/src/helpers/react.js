import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';
import { propTypes as mobxPropTypes } from 'mobx-react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import styled, { css } from 'styled-components';
import { idType } from 'shared';

export {
  React, cn, ReactDOM, mobxPropTypes, PropTypes,
  invariant, styled, css, idType,
};
export { observable, action, computed } from 'mobx';
export { observer, inject, Provider } from 'mobx-react';
