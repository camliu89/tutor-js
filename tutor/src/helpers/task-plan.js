import { isEmpty, get, pick, first, every, reduce, map } from 'lodash';
import moment from 'moment';
import Time from '../models/time';

export default {

  // returns the opens_at and due_at dates combined if they're identical,
  // separated by target_id if not
  dates(plan, options) {
    if (options == null) { options = {}; }
    if (isEmpty( get(plan, 'tasking_plans' ))) { return {}; }
    const pickFn = options.only ? tp => pick(tp, options.only) : tp => pick(tp, 'opens_at', 'due_at');

    const firstTasking = first( plan.tasking_plans );
    const isIdentical = every(plan.tasking_plans, function(tp) {
      if (options.only != null) {
        return tp[options.only] === firstTasking[options.only];
      } else {
        return (tp.opens_at === firstTasking.opens_at) && (tp.due_at === firstTasking.due_at);
      }
    });
    if (isIdentical) {
      return { all: pickFn(firstTasking) };
    } else {
      return reduce( plan.tasking_plans, function(memo, tp) {
        memo[tp.target_id] = pickFn(tp);
        return memo;
      }
      , {});
    }
  },

  calendarParams(course) {
    let date;
    const dueAtRange = course
      .teacherTaskPlans.array
      .map(plan => plan.dateRanges.due.start)
      .sort();

    if (dueAtRange.length) {
      date = dueAtRange[0];
    } else {
      date = Time.now;
    }
    return {
      to: 'calendarByDate',
      params: {
        courseId: course.id,
        date: moment(date).format('YYYY-MM-DD'),
      },
    };
  },

  earliestDueDate(plan) {
    const dates = map(get(plan, 'tasking_plans'), 'due_at');
    return first(dates.sort()) || '';
  },

};
