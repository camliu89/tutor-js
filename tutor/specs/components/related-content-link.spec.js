import { Factory } from '../helpers';
import ChapterSection from '../../src/models/chapter-section';
import RelatedContentLink from '../../src/components/related-content-link';
import RelatedContent from '../../src/models/related-content';

describe('Related Content Link', () => {

  it('renders and matches snapshot', () => {
    const content = [
      new RelatedContent({
        title: 'Intro', chapter_section: new ChapterSection([1,0])
      }),
      new RelatedContent({
        title: 'First', chapter_section: new ChapterSection([1,1])
      }),
    ];
    expect.snapshot(
      <RelatedContentLink course={Factory.course()} content={content} />
    ).toMatchSnapshot();
  });

});
