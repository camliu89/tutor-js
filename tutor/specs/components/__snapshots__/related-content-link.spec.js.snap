// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related Content Link renders and matches snapshot 1`] = `
.c0 .chapter-section {
  font-weight: normal;
  color: inherit;
}

.c1 {
  margin-left: 0.5rem;
}

<div
  className="c0"
>
  <span
    className="preamble"
  >
    Comes from 
  </span>
  <a
    aria-label="Browse course reference book"
    className="browse-the-book"
    href="/book/1/section/1"
    onClick={[Function]}
    tabIndex={-1}
    target="_blank"
  >
    <span
      className="part"
    >
      <span
        className="chapter-section"
        data-chapter-section="1.0"
      >
        1
      </span>
      <span
        className="c1"
      >
        Intro
      </span>
    </span>
  </a>
  <a
    aria-label="Browse course reference book"
    className="browse-the-book"
    href="/book/1/section/1.1"
    onClick={[Function]}
    tabIndex={-1}
    target="_blank"
  >
    <span
      className="part"
    >
      <span
        className="chapter-section"
        data-chapter-section="1.1"
      >
        1.1
      </span>
      <span
        className="c1"
      >
        First
      </span>
    </span>
  </a>
</div>
`;
